xla/service/gpu/runtime/topk_kernel.cu.h

```diff
@@ -146,105 +98,116 @@ __device__ __forceinline__ int Idx(int i) {
       for (int j = i + 1; j < K; j++) {
         KVT ti = tmp[i];
         KVT tj = tmp[j];
-        bool cmp = Trait::Gt(ti, tj);
+        bool cmp = ti > tj;
         tmp[i] = cmp ? ti : tj;
         tmp[j] = cmp ? tj : ti;
       }
     }
+    constexpr uint32_t WarpSize = WAVEFRONT_SIZE;
 
     for (int idx = K; idx < n; idx++) {
-      KVT kv(key[Idx(idx)], Idx(idx));
+      KVT kv{key[Idx(idx)], Idx(idx)};
       Push(tmp, kv);
     }
+    Reduce(tmp, WarpSize);
 
-    Reduce(tmp, 32);
-
-    if (threadIdx.x % 32 != 0) return;
-    int warp_id = threadIdx.x / 32;
+    if (threadIdx.x % WarpSize != 0) return;
+    int warp_id = threadIdx.x / WarpSize;
+#pragma unroll    
     for (int i = 0; i < K; i++) {
-      buffer_[i * 32 + warp_id] = tmp[i];
+      buffer_[i * WarpSize + warp_id] = tmp[i];
     }
   }
 
   // Merge the per-warp topks into a single topk. The final data is written to
-  // `keys` and `values`
-  __device__ void MergeTopKs(KT* keys, uint32_t* values) {
+  // `keys` and `idxs`
+  __device__ void MergeTopKs(KT *keys, uint32_t *idxs) 
+  {
+    constexpr uint32_t WarpSize = WAVEFRONT_SIZE;
     KVT tmp[K];
     // We only use one warp for this step.
-    if (threadIdx.x / 32 != 0) return;
+    if (threadIdx.x >= WarpSize) return;
     __syncthreads();
 #pragma unroll
-    for (int i = 0; i < K; i++) tmp[i] = buffer_[i * 32 + threadIdx.x];
-    Reduce(tmp, blockDim.x / 32);
+    for (int i = 0; i < K; i++) {
+      tmp[i] = buffer_[i * WarpSize + threadIdx.x];
+    }
+    Reduce(tmp, blockDim.x / WarpSize);
     if (threadIdx.x != 0) return;
     for (int i = 0; i < num_outputs_; ++i) {
-      tmp[i].Write(&keys[i], &values[i]);
+      keys[i] = tmp[i].key;
+      idxs[i] = tmp[i].idx;
     }
   }
 
   // Merge `tmp` (a reverse-sorted array) from (0, `num_lanes`) lanes. The
   // resulting array is stored in the tmp array of lane 0. For all other lanes,
   // `tmp` is unspecified after this function is called.
-  __device__ __forceinline__ void Reduce(KVT tmp[K], int num_lanes) {
-    int lane_id = threadIdx.x % 32;
+  __device__ FORCEINLINE void Reduce(KVT tmp[K], int num_lanes) 
+  {
+    constexpr uint32_t WarpSize = WAVEFRONT_SIZE;
+    int lane_id = threadIdx.x % WarpSize;
     for (int offset = num_lanes / 2; offset > 0; offset /= 2) {
 #pragma unroll
       for (int i = 0; i < K; i++) {
-        KVT kv = tmp[i].ShuffleDown(offset);
+        KVT kv = gpuShuffle< ShflType::Down >(tmp[i], offset);
         if (lane_id >= offset) continue;
         Push(tmp, kv);
       }
     }
   }
 
-  // Given a K-array of previously reverse-sorted KVTs, add kv to to it and
+  // Given a K-array of previously reverse-sorted KVTs, add kv to it and
   // remove the smallest element of the resulting array. Preserves the sorted
   // order of `tmp`.
-  static __device__ __forceinline__ bool Push(KVT tmp[K], const KVT& kv) {
-    if (Trait::Gt(tmp[K - 1], kv)) return false;
-    tmp[K - 1] = kv;
-    if constexpr (K >= 2) {
```

Comment:
@majnemer gave an internal review comment and asked why can this line be removed?
According to my understanding, the loop bound check will do the same thing, but wanted to make sure I understand it correctly.

Reply:
I guess you are referring to 'if constexpr()' statement. Yes, this part is obsolete because K is a compile-time constant and the loop below is getting unrolled anyway, so it will be completely optimized out for K = 1:
```
#pragma unroll
for (int i = (int)K - 2; i >= 0; --i) {
      if (tmp[i] > kv) break;
      // Swap
      KVT t = tmp[i];
      tmp[i] = tmp[i + 1];
      tmp[i + 1] = t;
    }
```